system_test/helpers/aest_docker.erl

```diff
@@ -0,0 +1,260 @@
+-module(aest_docker).
+
+%=== EXPORTS ===================================================================
+
+%% API exports
+-export([start/1]).
+-export([stop/1]).
+-export([setup_node/2]).
+-export([delete_node/1]).
+-export([start_node/1]).
+-export([stop_node/1, stop_node/2]).
+-export([kill_node/1]).
+-export([get_service_address/2]).
+
+%=== MACROS ====================================================================
+
+-define(CONFIG_FILE_TEMPLATE, "epoch.yaml.mustache").
+-define(EPOCH_CONFIG_FILE, "/home/epoch/epoch.yaml").
+-define(EPOCH_LOG_FOLDER, "/home/epoch/node/log").
+-define(EPOCH_MINE_RATE, 1000).
+-define(EXT_HTTP_PORT, 3013).
+-define(INT_HTTP_PORT, 3113).
+-define(INT_WS_PORT, 3114).
+
+%=== TYPES =====================================================================
+
+-type log_fun() :: fun((io:format(), list()) -> ok) | undefined.
+-type test_uid() :: binary() | undefined.
+-type service_label() :: ext_http | int_http | int_ws.
+
+%% State of the docker backend
+-type backend_state() :: #{
+    postfix := binary(),        % A unique postfix to add to container names.
+    log_fun := log_fun(),       % Function to use for logging.
+    data_dir := binary(),       % The directory where the templates can be found.
+    temp_dir := binary(),       % A temporary directory that can be used to generate
+                                % configuration files and save the log files.
+    net_id := binary()          % Docker network identifier
+}.
+
+%% Node specification
+-type node_spec() :: #{
+    name := atom(),
+    % Names or URLs of the peer nodes
+    peers := [atom() | binary()],
+    source := {pull, binary()}  % Source of the node image
+}.
+
+
+%% State of a node
+-type node_state() :: #{
+    spec := map(),              % Backup of the spec used when adding the node
+    log_fun := log_fun(),       % Function to use for logging
+    hostname := atom(),         % Hostname of the container running the node
+    exposed_ports := #{service_label() => pos_integer()},
+    local_ports := #{service_label() => pos_integer()}
+}.
+
+
+%=== GENERIC API FUNCTIONS =====================================================
+
+-spec start(Options) -> State
+    when Options :: #{test_id => test_uid(),
```

Comment:
`Options` should probably an `options()` type?

Reply:
It could, but it is only for the stop function so you would have to create a `start_option()` type then...

Reply:
`start_options()` is fine.